### ISSUE TITLE
Cannot save new file/or open exiting files (Rodin 3.4)

### ISSUE THREAD

**Issue by bmaderbacher** [hhu-stups/camille#26]
### Problem
I created a new project in a fresh Rodin installation and added a new context to it.
As an example I used the array context from [http://wiki.event-b.org/index.php/TextEditor_EBNF](http://wiki.event-b.org/index.php/TextEditor_EBNF).
When I try to safe I get the error message displayed below.
I get the a similar error on opening an exiting context or machine file, created with the rodin editor.

### Environment: 
- Rodin 3.4 for linux (rodin-3.4.0.201802230927-6980ca1-linux.gtk.x86_64)
- OpenJDK 8 (8.u162)
- Camille nightly (3.2.0.201803161304)
- no additional plugins

### Error message:
`Save Failed
Couldn't add in target because its parent hasn't been merged yet: MERGING LEFT DELETE org.eclipse.emf.compare.internal.spec.ReferenceChangeSpec{reference=BinaryOperator.left,value=IdentifierExpression@71135794 n,parentMatch=org.eclipse.emf.compare.internal.spec.MatchSpec{left=<null>,right=BelongPredicate@35451ba6 http://emf.eventb.org/models/core/formulas::BelongPredicate::array.axm1._OuTcwDHhEeiosJ6emmGu5g,origin=<null>,#differences=2,#submatches=0},match of value=org.eclipse.emf.compare.internal.spec.MatchSpec{left=<null>,right=IdentifierExpression@71135794 n,origin=<null>,#differences=0,#submatches=0}}`

This project looks really nice and I hope someone can help me to get it to run.
If you need additional information I'll be happy to provide them.
Yours,
Benedikt

**Comment by wysiib** [comment]
Apparently, something changed in the EMF Synchronisation Framework between Rodin 3.3 and Rodin 3.4. So far, I do not no what changed exactly.

**Comment by weetmuts** [comment]
Camille is such a greate editor that I cannot work in Rodin without it. This means that this bug blocks me from upgrading to Rodin 3.4.... If there is something I can do to help out, please let me know.

**Comment by weetmuts** [comment]
If you could explain which part of the Camille code is responsible for the
merge. Like filename and function names, that would help a lot. I have
already built Rodin and Camille.

//Fredrik

Den ons 11 apr. 2018 16:02Sebastian Krings <notifications@github.com> skrev:

**Comment by cfsnook** [comment]
I just had a quick try of Camille in Rodin 3.4.  The problem looks like it might be only with elements that have formula : expressions and predicates. Adding variables, events, parameters all works.. invariants, guards, actions cannot be saved. Maybe something to do with the merging of formulas or  formulas meta-model?

**Comment by cfsnook** [comment]
Maybe it is trying to convert the EMF model of the formula to a string before the parent EMF element (invariant, guard etc) has been merged/created.  (Only Camille uses an EMF model for the formula, it has to be reduced to a string (predicate/expression) when Camille saves).

**Comment by bodeveix** [comment]
I have looked at this problem and it seems to be fixed even if the fix is not very clean. I have modified PersistenceHelper to call directly the EventBMerger and to catch potential exceptions when applying a diff: this update avoids exceptions on Save. Finally, no exceptions seem to occur thanks to the EventBMerger update. In EventBMerger, I have used the AttributeChangeMerger and ReferenceChangeMerger of the library. I have also tried to merge the location information. Line information seems correct but sub-expressions are not well marked. I think it would need to find the index of the change in a multi-valued attribute.
It seems the isMergedFor method could simply return true. I have kept a part of the previous code. It was supposed to avoid erasing information not managed by camille, but it seems there is nothing to do for that.
Another problem I have tried to fix is that saving does not always launch workspace update.  Adding
			```resource.setModified(true);
			resource.eSetDeliver(true);```
in PersistenceHelper seems to reduce the occurrences of the problem.

I attach the two updated files and the .jar that can be used to replace the one in Rodin 3.4 (with identical version number).

Note that I have not made extensive tests. Thus, use it at your own risk...

[PersistenceHelper.java.zip](https://github.com/hhu-stups/camille/files/2542076/PersistenceHelper.java.zip)
[EventBMerger.java.zip](https://github.com/hhu-stups/camille/files/2542079/EventBMerger.java.zip)
[org.eventb.texttools.zip](https://github.com/hhu-stups/camille/files/2542080/org.eventb.texttools.zip)

**Comment by leuschel** [comment]
Many thanks for your efforts!
The fix seems to work on my Rodin 3.4 installation.

I have pushed the changes to the github repository.
Unfortunately building has failed on Travis:

$ gem install travis-custom-deploy
YAML safe loading is not available. Please upgrade psych to a version that supports safe loading (>= 2.0).
ERROR:  Error installing travis-custom-deploy:
	The last version of net-ssh (>= 2.6.5) to support your Ruby & RubyGems was 4.2.0. Try installing it with `gem install net-ssh -v 4.2.0` and then running the current command again
	net-ssh requires Ruby version >= 2.2.6. The current ruby version is 2.0.0.
The command "gem install travis-custom-deploy" failed and exited with 1 during .

Greetings,
Michael

**Comment by leuschel** [comment]
Your changes have now been successfully integrated into Camille and are available from the Camille update site.

Concerning the workspace update: I sometimes have the problem that saving a changed model works in the sense that e.g. other editors like the Rodin Editor see the changes, but the POs are not updated.
Sometimes it is very hard or even impossible, even using clean, to get Rodin to update the POs.
This looks more like a Rodin bug than a Camille issue.

**Comment by cfsnook** [comment]
Just a comment re: 
" I have kept a part of the previous code. It was supposed to avoid erasing information not managed by camille, but it seems there is nothing to do for that."
This is important for me... Camille does not load iUML-B diagrammatic models as text so without this code it would see my diagrams as something to be deleted when it came to merge.

**Comment by bodeveix** [comment]
Yes, I can save the changes as for any Event-B model. If I close and re-open the Event-B model, I see the changes. I get the same behaviour as with the Rodin editor.

**Comment by cfsnook** [comment]
I think maybe you tried the old UML-B (which should not be affected by Camille because it just regenerates the whole Event-B project and overwrites anything you edit with Camille).

 iUML-B is different because it stores the state-machine model inside the machine that Camille is editing.
I just tried it with iUML-B state-machines and Camille did not preserve eOpposite references in a state-machine transition.
 i.e. it preserved all the elements in the state-machine and the references from state to  their incoming and outgoing transitions but the transition lost its references to source and target states and became invalid. 
However, I think we have seen this before so it may have been the case before your edits.

[NewProject_201811032126.tar.gz](https://github.com/hhu-stups/camille/files/2545360/NewProject_201811032126.tar.gz)

**Comment by bodeveix** [comment]
I have made new changes to make Camille compatible with the new Rodin core. The iUML-B statemachine diagram problem is also solved. I attach the new project sources files. I have incremented the version number.
[org.eventb.texteditor.src.zip](https://github.com/hhu-stups/camille/files/2565757/org.eventb.texteditor.src.zip)

**Comment by bodeveix** [comment]
I have noticed an error linked to the object comparator. As a consequence, values deleted in Camille from multi-valued references (for example "sees") were not delete in the model. I attach the new project source file.
[org.eventb.texteditor.src.zip](https://github.com/hhu-stups/camille/files/2592664/org.eventb.texteditor.src.zip)
